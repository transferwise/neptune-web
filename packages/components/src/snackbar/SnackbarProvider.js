import React, { Component } from 'react';
import PropTypes from 'prop-types';

import SnackbarPortal from './Snackbar';

export const SnackbarContext = React.createContext();
export const SnackbarConsumer = SnackbarContext.Consumer;

class SnackbarProvider extends Component {
  constructor() {
    super();
    this.state = {
      text: '',
      timestamp: 0,
    };
  }

  create = ({ action, text, theme }) => {
    this.setState({ action, text, theme, timestamp: Date.now() });
  };

  render() {
    const { action, text, theme, timestamp } = this.state;

    return (
      <SnackbarContext.Provider
        value={{
          createSnackbar: this.create,
        }}
      >
        <SnackbarPortal
          action={action}
          text={text}
          timestamp={timestamp}
          timeout={this.props.timeout}
          theme={theme}
        />
        {this.props.children}
      </SnackbarContext.Provider>
    );
  }
}

SnackbarProvider.propTypes = {
  children: PropTypes.node.isRequired,
  timeout: PropTypes.number,
};

SnackbarProvider.defaultProps = {
  timeout: 4500,
};

export default SnackbarProvider;
