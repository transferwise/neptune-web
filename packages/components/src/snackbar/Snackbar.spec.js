import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import SnackbarAppendingToBody, { Snackbar, CSS_TRANSITION_DURATION } from './Snackbar';
import SnackbarProvider, { SnackbarConsumer } from './SnackbarProvider';

describe('Snackbar', () => {
  const timeout = 1000;
  let props;
  let component;
  let snackbar;
  let buttonTrigger;

  beforeEach(() => {
    jest.useFakeTimers();
    props = {
      text: 'hello, i am a snack',
      timestamp: Date.now(),
      timeout,
    };

    const Container = () => (
      <SnackbarConsumer>
        {({ createSnackbar }) => (
          <button type="button" className="button-trigger" onClick={() => createSnackbar(props)}>
            Trigger
          </button>
        )}
      </SnackbarConsumer>
    );

    component = mount(
      <SnackbarProvider timeout={timeout}>
        <Container />
      </SnackbarProvider>,
    );
    snackbar = () => component.find(Snackbar);
    buttonTrigger = () => component.find('.button-trigger');
  });

  it('is appended to body', async () => {
    const createPortal = jest.spyOn(ReactDOM, 'createPortal');
    expect(createPortal).not.toHaveBeenCalled();
    mount(<SnackbarAppendingToBody {...props} />);

    expect(createPortal).toBeCalledTimes(1);
    /** Using toBeCalledWith was not matching properly */
    const [comp, body] = ReactDOM.createPortal.mock.calls[0];
    expect(comp).toMatchObject(snackbar());
    expect(body).toMatchObject(document.body);
    jest.clearAllMocks();
  });

  it('accepts element as text', async () => {
    const componentWithNode = mount(
      <SnackbarProvider timeout={timeout}>
        <SnackbarConsumer>
          {({ createSnackbar }) => (
            <button
              type="button"
              className="button-trigger"
              onClick={() => createSnackbar({ text: <span>test</span> })}
            >
              Trigger
            </button>
          )}
        </SnackbarConsumer>
      </SnackbarProvider>,
    );
    const snackbarWithNode = () => componentWithNode.find(Snackbar);
    componentWithNode.find('.button-trigger').simulate('click');
    expect(snackbarWithNode().text()).toContain('test');

    expect(snackbarWithNode().html()).toMatchSnapshot();
  });

  it('displays a single snack for the given duration with the given text', async () => {
    expect(snackbar().text()).not.toContain(props.text);
    buttonTrigger().simulate('click');
    buttonTrigger().simulate('click');
    buttonTrigger().simulate('click');
    buttonTrigger().simulate('click');
    expect(snackbar().text()).toContain(props.text);
    expect(snackbar().length).toBe(1);

    jest.advanceTimersByTime(timeout + CSS_TRANSITION_DURATION + 500);

    expect(snackbar().text()).not.toContain(props.text);
  });

  it('adjusts the theme if passed', async () => {
    const componentWithNode = mount(
      <SnackbarProvider timeout={timeout}>
        <SnackbarConsumer>
          {({ createSnackbar }) => (
            <button
              type="button"
              className="button-trigger"
              onClick={() => createSnackbar({ text: <span>test</span>, theme: 'dark' })}
            >
              Trigger
            </button>
          )}
        </SnackbarConsumer>
      </SnackbarProvider>,
    );
    const snackbarWithNode = () => componentWithNode.find(Snackbar);
    componentWithNode.find('.button-trigger').simulate('click');
    expect(snackbarWithNode().html()).toMatchSnapshot();
  });
});
