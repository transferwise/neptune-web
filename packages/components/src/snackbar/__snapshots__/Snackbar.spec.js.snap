// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar accepts element as text 1`] = `"<div class=\\"snackbar\\"><span class=\\"snackbar__text snackbar__text--light snackbar__text-container-enter snackbar__text-container-enter-active\\"><span>test</span></span></div>"`;

exports[`Snackbar adjusts the theme if passed 1`] = `"<div class=\\"snackbar\\"><span class=\\"snackbar__text snackbar__text--dark snackbar__text-container-enter snackbar__text-container-enter-active\\"><span>test</span></span></div>"`;
